bin/types-auto-pr.ts

```diff
@@ -0,0 +1,87 @@
+// Run CLI using @expo/spawn-async
+
+/*
+Clone DT fork:
+
+git clone --depth=1 https://github.com/Maxim-Mazurok/DefinitelyTyped --no-single-branch # clone only last commit for every branch
+
+*/
+
+/*
+Rebase master:
+
+cd DefinitelyTyped
+git remote add upstream https://github.com/DefinitelyTyped/DefinitelyTyped
+DATE_SINCE=$(git log -1 --format="%cd" --date=unix) # get time of the last commit to origin
+git fetch upstream master --shallow-since=$DATE_SINCE # update master from upstream since last commit to origin (to have all required history for push)
+git reset --hard upstream/master # reset fork to upstream
+git push # actually update master
+
+*/
+
+/*
+Copy "types" branch of generator fork
+curl https://codeload.github.com/Maxim-Mazurok/google-api-typings-generator/tar.gz/types | tar xvz -
+rm -f google-api-typings-generator-types/.gitkeep # TODO: find a better way to not include this file
+rsync -a google-api-typings-generator-types/ types/ # TODO: handle deleted files
+rm -rf google-api-typings-generator-types # remove original folder after copying
+git add -A # add all modified/added files to git
+git commit -m "tmp" # commit changes so that test runner can detect what changed
+npm i # install required packages to run tests (no `npm ci` because https://github.com/DefinitelyTyped/DefinitelyTyped/pull/30476#issuecomment-593053174)
+npm test # run DT test runner on changed types # TODO: might want to delete non-gapi types to speed up parsing phase
+git reset HEAD^ --quiet # undo last commit
+# use parse-git-status to find all changed types
+
+For each type without branch:
+git checkout -b gapi.client.TYPE_NAME
+git add types/gapi.client.TYPE_NAME
+git commit -m "automatic types update @ {yyyy-mm-dd hh:mm}"
+*/
+
+import path from 'path';
+import parseGitStatus from 'parse-git-status';
+import spawnAsync from '@expo/spawn-async';
+import {typingsPrefix} from "../src/app";
+
+class TypesAutoPr {
+  readonly #cwd: string;
+
+  constructor(cwd?: string) {
+    this.#cwd = cwd || path.resolve(__dirname, '../DefinitelyTyped');
+  }
+
+  getChangedTypes = async (): Promise<string[]> => {
+    const {stdout} = await spawnAsync('git', ['status', '--porcelain', '-z'], {
+      cwd: this.#cwd
+    });
+    const status = parseGitStatus(stdout);
+    return [...new Set(status.map(x => path.basename(path.parse(x.to).dir)))].filter(x => x.indexOf(typingsPrefix) === 0)
+  };
```

Comment:
Is it enough to just check if these changed or do we also want to ensure the revision number _increased_?

Reply:
Thanks for taking a look! Yes, that's a valid concern. The problem is that we don't know how revisions actually work. I saw revision number increasing and _decreasing_ also. I assume that Google may not only release new revisions, but also rollback to older revisions. It would be great to ask somebody from Google regarding this. Maybe the reason why I saw decreasing revision numbers is because of soft-releases but we can't be sure.

Anyway, whatever logic we will use for revisions (use latest revision, or latest fetched JSON), it should be applied on types generation/publication step, I think. This script should just use whatever is in `types` branch. Otherwise, we might end up checking the whole commits history to figure out which revision is the latest. It would be much easier if we just check during generation/publication to `types` branch is we're publishing revision which is never than current revision. So it will never decrease.

Makes sense?

Reply:
Yeah, this sounds good. We can take another pass at the build stage to only publish new revisions.